### PR TITLE
Only allow one client to produce a symbol tree info at a time.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Refresher:  SymbolTreeInfo is the type we keep around that acts as an index (by name) of all the top level symbols in a source compilation or dll.  It is used by numerous IDE features to quickly go from Name-to-Symbol (for example with features like Add-Using).  SymbolTreeInfo keeps both exact indices (i.e. the searched name must exactly match the symbol name) as well as fuzzy indices (i.e. the searched name needs to be 'close enough' to the symbol name).  

This change addresses two problems we were having with the current system around SymbolTreeInfo.

First:
The current implementation of SymbolTreeInfo does no synchronization when clients call in to create one.  This means if we have N threads trying to get the same SymbolTreeInfo we'll perform the computation for it N times, keeping only the last result.  This is pretty wasteful.  And, in my own personal testing, was happening a lot.

Now, we only allow a single SymbolTreeInfo to be created at a time, so that we don't waste any CPU producing and then throwing away work.

--

Second:
We were caching SymbolTreeInfos in a way that was causing them to be created and thrown out far more often than we would have liked.  For metadata dlls we want the lifetime of the SymbolTreeInfo to be as long as the metadata we've loaded for them.  In the IDE we load this metadata once and keep it around as long as the file hasn't changed on disk.  The introduction of the Metadata ID works to allow us to represent that lifetime, and associate our indices appropriately.  i.e. as long as a MetadataReference is still producing and caching (in the IDE case) Metadata with the same ID, then our indices for that dll will stay alive.  When the IDE dumps this metadata, then we'll dump the index as well.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @tmat This is what i'm currently thinking about as an approach to the metadata caching issue we discussed today.

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/mac/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_mac_dbg_unit32/2462/
// Retest reason: timeout

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
Tagging @dpoeschl in case you're interested in looking at this.

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by mattwar** [comment]
Metadata Id should be its own type, since you are in fact making new public API anyway to access this value, and having Id types is the pattern that the rest of Roslyn uses.

**Comment by tmat** [comment]
If we have a pattern we use for this then I'm fine with following the pattern.

**Comment by tmat** [comment]
Perhaps the type can be nested (Metadata.Id) to avoid clutter?

**Comment by mattwar** [comment]
Nope, it has to be MetadataId.

**Comment by CyrusNajmabadi** [comment]
Ok.  will revert it back to MetadataId.

**Comment by tmat** [comment]
OK.

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/win/dbg/eta please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_dbg_eta/3002/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Review IDE bits please.
@dotnet/roslyn-compiler Review compiler bits please.

File: src/Compilers/Core/Portable/MetadataReference/Metadata.cs [hunk 1]
```diff
@@ -6,15 +6,34 @@
...
 {
     /// <summary>
+    /// An Id that can be used to identify a metadata instance.  If two metadata instances 
+    /// have the same id then they are guaranteed to have the same content.  If two metadata
+    /// instances have different ids, then the contents may or may not be the same.  As such,
+    /// the id is useful as a key in a cache when a client wants to share data for a metadata
+    /// reference as long as it has not changed.
+    /// </summary>
+    public sealed class MetadataId
+    {
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: Feels strange to have a type with no members.  Is there an expectation that we'll ever have members on this type?  If not I'm curious what the advantage is of using this type over simply `object`.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: Hey Jared, see the above convo with Matt/Tomas.  My personal main objection over having this be ```object``` is that it gets *super* easy to accidentally use the wrong value as the compiler/language gives you zero type safety.  For example, there is code that looks like this in this pr:  ```base(..., id: reference.Id)```.  If the 'id' were typed as ```object```, it would have been trivial to accidently mistype that as:  ```base(..., id: reference)``` instead.  In that case the compiler would have given me no information that i did anything wrong.  However, because it is typed as MetadataId, i must provide an actual value typed that way.

Similarly, in the code that consumes this later, i have a ```ConditionalWeakTable<MetadataId, ...>```.  In this form the compiler ensures that i'm looking up/storing with actual MetadataIds.  If it were typed as object it would have been trivial to accidently write ```map.GetOrAdd(ref.GetMetadata(), ...)``` instead of ```map.GetOrAdd(ref.GetMetadata().Id, ...)```.

So, in other words, even though it is an empty type, it still provides real type safety in the code.  Something i would like to keep around.  

So personally, i would like it to be typed.  It reduces the chance of errors.  And, if we want to add state to it (perhaps for debugging purposes), now it is trivial to do so.

Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: Okay.  Still wish there was something that could be added to at least give the type instances some debuggability value but I understand the strong type argument.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: Well, these are all made from a PortableExecutableReference.  And those have a FilePath.  So i think we could use the filepath, plus a monotonically increasing integer for that purpose.  I can make that change if you'd like that :)

Comment (jaredpar) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: I'll defer to other peoples opinions here.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: Actually, i take it back.  Due to how these things are created (esp. from random streams in memory), it's not altogether easy to provide such a thing.

**Comment by AlekseyTs** [comment]
Is there a description of the problem that is addressed by this change? What is "symbol tree info"? Why only one client should be allowed to produce it?

**Comment by CyrusNajmabadi** [comment]
Hey @AlekseyTs.  Sorry about that.  I've included a writeup now at the start of the PR to better explain what's going on.

File: src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs [hunk 2]
```diff
@@ -146,18 +146,30 @@ internal sealed override MetadataReference WithPropertiesImplReturningMetadataRe
...
         }
 
+        /// <summary>
+        /// Returns a fresh copy of the <see cref="Metadata"/> referenced by this <see cref="MetadataReference"/>.
+        /// The instance will not be be the exact same one used by the compiler.  As such, disposing of the
+        /// metadata returned by this will have no effect on the metadata used by existing compilation instances.
+        /// </summary>
+        /// <exception cref="IOException"/>
+        /// <exception cref="BadImageFormatException"/>
+        public Metadata GetMetadata()
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: Perhaps we should use more descriptive name, for example "GetSafeDisposableMetadataCopy", "GetOwnMetadataCopy", "GetMetadataCopy", etc. and keep the name of the original API as is.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: @AlekseyTs I'm fine with that.  @tmat Is there a particular name you think would be appropriate here to properly convey this?

File: src/Compilers/Core/Portable/MetadataReference/UnresolvedMetadataReference.cs [hunk 3]
```diff
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReference/UnresolvedMetadataReference.cs]: It looks like this is the only change in this file, please revert.

File: src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: Also, it looks like someone else can still dispose underlying data for the copy. That should probably be reflected in the comment. Basically you own the returned object, but not the underlying data.

**Comment by AlekseyTs** [comment]
In context of different compilations, the same metadata reference can have different corresponding sets of symbols. Or, if a new reference is added to the same compilation, it can change symbols used for previously referenced metadata/compilations. Bottom line - the same metadata id doesn't guarantee the same set of symbols. Does this make any difference for SymbolTreeInfo? If so, does implementation account for that?

**Comment by AlekseyTs** [comment]
Compiler side of the change LGTM, modulo the naming and the comment for the new API on PortableExecutableReference.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs The cases where the set of symbols might change are indeed interesting.  My gut tells me we don't need to care too deeply about that for our scenarios.  All that generally means is that in some rare cases add-using may not find something.  But nearly all the time with nearly all types things will still work fine for the user.

File: src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: Note: because of all these subtle aspects, i'm leaning heavily to only wanting to expose the ID directly off the metadata reference.  i.e. ```PortableExecutableReference.GetCurrentMetadataId()```.  What do you guys think?

Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: Isn't that a race? How do you synchronize that with the GC potentially getting rid of stuff?

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: It's a race in the sense that at one point we may say: yes you can reuse the index.  But all that means is that we may or may not get precisely correct results for a lookup call (if we happen to lookup *just as metadata is changing on disk*)

But the very next time we call in, the Id will be different and we'll compute the index.  To me, this seems like a totally acceptable bit of behavior...

Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/MetadataReference/PortableExecutableReference.cs]: Sure, it's probably fine for your scenario. It might _not_ be fine for the next scenario that wants this API. Designing APIs that have races that "don't matter" for the first use case probably isn't good API design. :smile:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski Well, it's like any sort of 'reference' API.  It tells you what it thinks now.  It's up to you to decide if that's ok.  Similar to any of our File apis.  I mean, I suppose we could revamp all of this to be more like a Capabilities API a-la midori, but that doesn't seem likely.  

Or, more specifically, do you have some sort of suggestion for what we should do here instead?

**Comment by jasonmalinowski** [comment]
Is moving the API to Compilation (some sort of GetMetadataId(reference)) at least ensure that once you have a compilation in hand you can get right thing?

**Comment by CyrusNajmabadi** [comment]
Ok.  Talking to Tomas, it might be possible.  Investigating.

**Comment by CyrusNajmabadi** [comment]
Ok all.  I've pushed another approach.  this time we expose the MetadataId through the IAssemblySymbol itself.  This means that we don't have the issue of 'raceyness' where you ask it off of the metadata reference itself.  Now if you have two IAssemblySymbols produced off the same underlying AssemblyMetadata, then you'll get the same ID, allowing us to reuse the appropriate indices.

One things i was unsure about was how to take care of RetargettingAssemblySymbols.  @AlekseyTs could you take a look and tell me if i did things in the appropriate fashion.  I have a feeling i may not have.

Thanks!

**Comment by tmat** [comment]
Could you squash your commits so that it's clear what's actually changed?

Could we instead of MetadataId expose Metadata off of IAsseblySymbol. That seems generally useful. If you just need Id you'd use Metadata.Id.

**Comment by CyrusNajmabadi** [comment]
Sure.  I assume you'd want a copy of the metadata returned, and not the actual metadata itself?

Though i'm a bit wary about exposing that.  Seems odd to expose something off an immutable API that may return different results depending on if it has been disposed or not.

**Comment by tmat** [comment]
Yes, the real, disposable Metadata shall not be exposed. 
All symbols have such property. If the host disposes the underlying metadata the corresponding symbols might AV (since they are lazily read from the metadata).

**Comment by CyrusNajmabadi** [comment]
Ah ok.  Will do then.

**Comment by AlekseyTs** [comment]
What about modules, they are backed by Metadata too?

**Comment by tmat** [comment]
Yes, modules should also expose Metadata.

**Comment by tmat** [comment]
In fact, assemblies should return AssemblyMetadata and modules ModuleMetadata

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
Ok, let me know what you guys think of the new approach.

**Comment by tmat** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by nguerrera** [comment]
To answer @tmat, this looks good and generally useful to me. :smiley: 

(I didn't study the subtleties about lifetime/copying/disposal.)

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide review appreciated.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 4]
```diff
@@ -24,6 +24,9 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     internal sealed class PEModule : IDisposable
     {
+        // We need to store reference for to keep the metadata alive while symbols have reference to PEModule.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: Typo "for to"?

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 5]
```diff
@@ -24,6 +24,9 @@ namespace Microsoft.CodeAnalysis
...
+        private readonly ModuleMetadata _owner;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: Why do we need this now, how did we manage without it so far and what has changed?

Comment (tmat) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: I think it's likely always been a problem but we don't work too much with standalone .netmodules so we haven't hit it. It's be good to add a dedicated test that creates a netmodule, only holds on the corresponding PE symbol (and not the metadata) calls GC.Collect and then accesses some members of the symbol.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: It is obvious we need it to implement GetNonDisposableMetadata, I am curious about the comment regarding live-time management.

File: src/Compilers/VisualBasic/Portable/Symbols/AssemblySymbol.vb [hunk 6]
```diff
@@ -79,6 +79,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         End Property
 
         ''' <summary>
+        ''' If this symbol represents a metadata assembly returns the underlying <see cref="AssemblyMetadata"/>.
+        ''' 
+        ''' Otherwise, this returns <code>null</code>.
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/AssemblySymbol.vb]: Nothing vs. null?

File: src/Compilers/VisualBasic/Portable/Symbols/ModuleSymbol.vb [hunk 7]
```diff
@@ -124,6 +124,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         End Property
 
         ''' <summary>
+        ''' If this symbol represents a metadata module returns the underlying <see cref="ModuleMetadata"/>.
+        ''' 
+        ''' Otherwise, this returns <code>null</code>.
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/ModuleSymbol.vb]: Nothing vs. null?

**Comment by AlekseyTs** [comment]
Compiler changes LGTM, modulo minor issues around comments.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs (hunk 4 above)
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: Should this be using XML doc comments?

File: src/Compilers/Core/Portable/MetadataReference/Metadata.cs [hunk 8]
```diff
@@ -6,15 +6,34 @@
...
+    /// <summary>
     /// Represents immutable assembly or module CLI metadata.
     /// </summary>
     public abstract class Metadata : IDisposable
     {
         internal readonly bool IsImageOwner;
 
-        internal Metadata(bool isImageOwner)
+        /// <summary>
+        /// The id for this metadata instance.  If two metadata instances have the same id, then
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/MetadataReference/Metadata.cs]: Casing of "id" here?

**Comment by jmarolf** [comment]
:+1: Pulled these changes down and confirmed that this fixes the perf issues I was seeing.  Is there any plans to add perf tests for this?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 9]
```diff
@@ -181,24 +180,33 @@ private int BinarySearch(string name)
...
 
         /// <summary>
         /// this gives you SymbolTreeInfo for a metadata
         /// </summary>
-        public static async Task<SymbolTreeInfo> GetInfoForAssemblyAsync(Solution solution, IAssemblySymbol assembly, string filePath, CancellationToken cancellationToken)
+        public static async Task<SymbolTreeInfo> TryGetInfoForAssemblyAsync(Solution solution, IAssemblySymbol assembly, PortableExecutableReference reference, CancellationToken cancellationToken)
         {
-            SymbolTreeInfo info;
-            if (s_assemblyInfos.TryGetValue(assembly, out info))
+            using (await s_assemblyInfosGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: Why the switch to serialize this work?  Do you no longer believe the old comment to be true?

File: src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs [hunk 10]
```diff
@@ -180,9 +180,9 @@ public void Test_SharedMetadataReferencesWithAliases()
                 Assert.NotNull(mdp2Sys3);
 
                 // all references to System.dll share the same metadata bytes
-                Assert.Same(mdp1Sys1, mdp1Sys2);
-                Assert.Same(mdp1Sys1, mdp2Sys1);
-                Assert.Same(mdp1Sys1, mdp2Sys3);
+                Assert.Same(mdp1Sys1.Id, mdp1Sys2.Id);
```
Comment (Pilchie) [inline src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs]: Why the change here.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs [hunk 11]
```diff
@@ -280,18 +280,23 @@ private static IEnumerable<ISymbol> TranslateNamespaces(List<ISymbol> symbols, C
...
 
             var result = new List<ISymbol>();
-            await AddDeclarationsAsync(solution, assembly, filePath, query, filter, result, cancellationToken).ConfigureAwait(false);
+            await AddDeclarationsAsync(solution, assembly, reference, query, filter, result, cancellationToken).ConfigureAwait(false);
             return result;
         }
 
         private static async Task AddDeclarationsAsync(
-            Solution solution, IAssemblySymbol assembly, string filePath, SearchQuery query, SymbolFilter filter, List<ISymbol> list, CancellationToken cancellationToken)
+            Solution solution, IAssemblySymbol assembly, PortableExecutableReference reference, SearchQuery query, SymbolFilter filter, List<ISymbol> list, CancellationToken cancellationToken)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs]: Should this be suffixed with opt or something since it can really very much be null?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs (hunk 9 above)
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: N/m.  I see the main description now. :blush:

Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: Still, should we consider allowing work to happen for each Id in parallel?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 12]
```diff
@@ -181,24 +180,33 @@ private int BinarySearch(string name)
...
             {
-                return info;
-            }
+                var metadata = assembly.GetMetadata();
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: Move this check outside the lock since it's not relying on locked state?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: We could.  Do you have a good suggestion on how best to achieve that?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs (hunk 12 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: sure.

File: src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs]: I could revert this.  Originally this was because of another change that caused us to return metadata *copies* (because we were now calling the public GetMetadata entrypoint).  These were necessarily *not* the same.   But the *ids* were the same.

**Comment by Pilchie** [comment]
IDE changes LGTM as long as @AlekseyTs and @tmat are happy with the metadata lifetime stuff.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs (hunk 11 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs]: sure.

**Comment by jasonmalinowski** [comment]
:+1:, with some high level concerns:

1. The compiler metadata APIs added seem generally reasonable, and I defer to @tmat and the compiler team's signoffs for that. (It looks like you already have that.) In particular, I understand and agree that the metadata can't be disposable from `IAssemblySymbol`, but it seems odd to me that it wouldn't throw -- calling assemblySymbol.GetMetadata().Dispose() will never work, and we just silently succeed there.
2. It's worrisome to me that we serialize generation of these, especially since we know they can be expensive. I'm OK with this change for now (since it's definitely better than what we currently have in master), but it's something we'll probably want to revisit.

Also, please rewrite history to make the commits meaningful. Given this went through several redesigns of the approach, we shouldn't have commits with all the abandoned approaches floating around with the serious ones.

File: src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs (hunk 10 above)
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs]: If the contract is only the Ids are equal, we should keep the tests. Otherwise, let's revert.

Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs]: The contract is that only the Ids are guaranteed.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: Ok. I've pushed a version that now allows computation across cores.

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
> It's worrisome to me that we serialize generation of these, especially since we know they can be expensive.

I've addressed this in my latest change.  Now we still limit computation for any single piece of metadata to a single core.  However, we can be generating trees for different metadatas simultaneously.

File: src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs [hunk 13]
```diff
@@ -40,7 +40,10 @@ internal sealed class PEAssembly
 
         private ThreeState _lazyDeclaresTheObjectClass;
 
-        // We need to store reference for to keep the metadata alive while symbols have reference to PEAssembly.
+        /// <summary>
+        /// We need to store reference to the assembly metadata to keep the metadata alive while 
+        /// symbols have reference to PEAssembly.
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs]: cref?

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 14]
```diff
@@ -24,6 +24,12 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     internal sealed class PEModule : IDisposable
     {
+        /// <summary>
+        /// We need to store reference to the module metadata to keep the metadata alive while 
+        /// symbols have reference to PEModule.
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: cref?
